[dnf5] Plugins architecture for libdnf

jrohel opened the pull request:
I designed the plugin architecture for libdnf 5.
The architecture supports plugins written in C++. However, it also supports loading plugins using plagins. Even plugins in languages ​​other than C ++.

`Base::load_plugins()` - Loads plugins (".so" files) from the directory specified in the `LIBDNF_PLUGIN_DIR` environment variable. If the environment variable is not set, plugins will not be loaded.

PR also includes a C++ plugin that loads Python plugins - python_plugins_loader. And one sample Python plugin.
The path to the directory with Python plugins is retrieved from the `LIBDNF_PYTHON_PLUGIN_DIR` environment variable. If this variable is not set, plugins are not loaded.

jrohel committed:
[plugin] API for C++ plugins

Defines the interface for plugins.

- The "libdnf::plugin::IPlugin" class.
  It will be the parent class (interface) of all plug-ins.

- "HookId" enumeration. List of all supported hooks.
  It will be updated gradually.

- Function to return the API version supported by the plugin.
  Same result as IPlugin::get_api_version(), but can be called
  without creating an IPlugin instance.
  libdnf::plugin::Version libdnf_plugin_get_api_version(void);

- Function to return the name of the plugin.
  Same result as IPlugin::get_name(), but can be called without
  creating an IPlugin instance.
  const char * libdnf_plugin_get_name(void);

- Function to return the plugin version.
  Same result as IPlugin::get_version(), but can be called
  without creating an IPlugin instance.
  libdnf::plugin::version libdnf_plugin_get_version(void);

- Functions for creating a new plugin instance:
  libdnf::plugin::IPlugin * libdnf_plugin_new_instance(libdnf::plugin::Version api_version);

- Function for deleting an existing plugin instance:
  void libdnf_plugin_delete_instance(libdnf::plugin::IPlugin * plugin_instance);

jrohel committed:
[utils] "Library" class - For working with shared libraries

It will be used to work with shared libraries containing the plugin.

jrohel committed:
[Base] Methods lock(), unlock(), get_locked_base()

Multiple "Base" instances can run in parallel in threads.
In some cases, only one "Base" instance can run at a time.

jrohel committed:
[plugin] Plugin access classes and Plugin manager

jrohel committed:
[plugin] Add SWIG bindings

jrohel committed:
[plugin] Add methods for working with plugins in Base

Added methods:
void add_plugin(plugin::IPlugin & iplugin_instance) - Adds a plugin to the base.

void load_plugins() - Loads plugins (".so" files) from the directory specified
        in the LIBDNF_PLUGIN_DIR environment variable. If the environment
        variable is not set, plugins will not be loaded.

plugin :: Plugins & get_plugins() - Returns the plugin manager.

jrohel committed:
[plugins] Prepare building of libdnf plugins

jrohel committed:
[plugins] Plugin for loading Python plugins

This libdnf plugin loads libdnf plugins written in Python.

The path to the directory with Python plugins is retrieved from
the LIBDNF_PYTHON_PLUGIN_DIR environment variable. If this variable is not set,
plugins are not loaded.

jrohel committed:
[plugins] Python test plugin for libdnf

Libdnf plugin written in Python. Processes the LOAD_CONFIG_FROM_FILE hook
and writes the current value of the "skip_if_unavailable" option to the log.

jrohel committed:
[microdnf] Load plugins

dmach commented:
Throws the following error:
```
terminate called after throwing an instance of 'std::filesystem::__cxx11::filesystem_error'
  what():  filesystem error: directory iterator cannot open directory: No such file or directory [libdnf-plugins/python_plugins_loader/]
Aborted (core dumped)
```
I'd also expect `libdnf-plugins/python_plugins_loader/` to be an absolute path.

dmach commented on libdnf-plugins/python_plugins_loader/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+endif()
+
+
+#if(WITH_BINDINGS)
+    if(NOT PYTHON_DESIRED)
+        find_package(PythonInterp REQUIRED)
+    elseif(${PYTHON_DESIRED} STREQUAL "2")
+        find_package(PythonInterp 2 EXACT REQUIRED)
+    elseif(${PYTHON_DESIRED} STREQUAL "3")
+        find_package(PythonInterp 3 EXACT REQUIRED)
+    elseif(EXISTS ${PYTHON_DESIRED})
+        set(PYTHON_EXECUTABLE ${PYTHON_DESIRED})
+        find_package(PythonInterp REQUIRED)
+    else()
+        message(FATAL_ERROR "Invalid PYTHON_DESIRED value: " ${PYTHON_DESIRED})
+    endif()
+    find_package(PythonLibs REQUIRED)
+#endif()
+
+find_package(PythonInstDir)
+include_directories(${PYTHON_INCLUDE_PATH})
```
please replace with the following lines:
```
find_package(Python3 REQUIRED COMPONENTS Interpreter Development)
include_directories(${Python3_INCLUDE_DIRS})
```

jrohel replied:
OK. Replaced

jrohel committed:
[plugins] Modify CMakeLists.txt and .spec files

The Libdnf plugin for loading plugins written in Python has its own build
option WITH_PYTHON_PLUGINS_LOADER.
Adds the package python3-libdnf-python-plugin-loader to the .spec file.

cgwalters commented on include/libdnf/plugin/iplugin.hpp:
```diff
@@ -0,0 +1,105 @@
+This file is part of microdnf: https://github.com/rpm-software-management/libdnf/
+
+Microdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
```
Did you mean this to be under the GPL?  The COPYING file claims libdnf is LGPLv2+.

jrohel replied:
No. I overlooked. Incorrectly used the license header for microdnf instead of libdnf.

cgwalters reviewed (commented):
Broadly speaking I think the "load dynamic language as plugin to C/C++ app" model hasn't been very successful in the industry, because basically there can only be one garbage collector in a single process.  This issue is why e.g. Microsoft spent so much time on .NET with the idea that there'd be a common runtime for multiple languages.

There will be at best unavoidable leaks if e.g. one plugin be in (c)Python and another in Go/JavaScript for example.

If the goal here is mostly "C++(Rust) or Python" that seems potentially viable at a fundamental level - but I'm still very skeptical of dynamic plugins for the OS updater/package manager use case.  It seems just much more reliable to e.g. ensure that a separate process can observe and react to the state of the system; perhaps via "exec callout" hooks.

But even then I think this should all be driven by concrete examples of a plugin and what they should do.

Conan-Kudo commented on libdnf/utils/library.cpp:
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of microdnf: https://github.com/rpm-software-management/libdnf/
+
+Microdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Microdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with microdnf.  If not, see <https://www.gnu.org/licenses/>.
```
This header is wrong. Please switch to the libdnf one (with LGPL license).

Conan-Kudo commented on libdnf/utils/library.hpp:
```diff
@@ -0,0 +1,72 @@
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with microdnf.  If not, see <https://www.gnu.org/licenses/>.
```
This header is wrong. Please switch to the libdnf one (with LGPL license).

Conan-Kudo commented on libdnf/plugin/plugins.cpp:
```diff
@@ -0,0 +1,140 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of microdnf: https://github.com/rpm-software-management/libdnf/
+
+Microdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Microdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with microdnf.  If not, see <https://www.gnu.org/licenses/>.
```
This header is wrong. Please switch to the libdnf one (with LGPL license).

Conan-Kudo commented on libdnf.spec:
```diff
@@ -294,6 +295,25 @@ Ruby bindings for the libdnf-cli library.
+%if %{with python_plugins_loader}
+%package -n python3-libdnf-python-plugins-loader
+Summary:        Libdnf plugin that allows loading Python plugins
+License:        GPLv2+
```
This should be LGPLv2

Conan-Kudo commented:
It is unclear to me whether we're spawning Python for each plugin or are able to use one Python interpreter instance for all of them. I would expect there to be a single interpreter here.

dmach commented:
> It is unclear to me whether we're spawning Python for each plugin or are able to use one Python interpreter instance for all of them. I would expect there to be a single interpreter here.

It's a single special C++ plugin that creates a bridge between C++ and Python and allows loading multiple Python plugins. There's a single interpreter loaded.

jrohel commented:
Rebased. I changed the license to LGPL.

Conan-Kudo commented on include/libdnf/plugin/iplugin.hpp:
```diff
@@ -0,0 +1,105 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on include/libdnf/plugin/plugins.hpp:
```diff
@@ -0,0 +1,159 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf-plugins/python_plugins_loader/plugin.py:
```diff
@@ -0,0 +1,56 @@
+#
+# Libdnf is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf-plugins/python_plugins_loader/python_plugins_loader.cpp:
```diff
@@ -0,0 +1,348 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf/plugin/plugins.cpp:
```diff
@@ -0,0 +1,140 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf/utils/library.cpp:
```diff
@@ -0,0 +1,52 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf/utils/library.hpp:
```diff
@@ -0,0 +1,72 @@
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
```
`version 2` -> `version 2.1`

Conan-Kudo commented on libdnf.spec:
```diff
@@ -294,6 +295,25 @@ Ruby bindings for the libdnf-cli library.
+%if %{with python_plugins_loader}
+%package -n python3-libdnf-python-plugins-loader
+Summary:        Libdnf plugin that allows loading Python plugins
+License:        LGPLv2.1+
```
`LGPLv2.1+` -> `LGPLv2+`

jrohel replied:
I copied the license header from another libdnf source file -> We need to check all the files.

dmach replied:
Yes, there's a problem in licenses in other files too.
I'll take care of them in another PR or push the changes directly.

dmach commented:
I've checked that the plugins work as expected (considering that this is an early code and remaining things as GIL/threading will be handled later) and the code builds on f33 and f35. Merging. Let's do more improvements in another PRs.
